STIPS/STIP-005.md

```diff
@@ -0,0 +1,1239 @@
+Simulate trade to get its real cost and discover the slippage & fees amount issuer will bear
+```solidity
+swapParams = {
+  baseToken: vBaseToken 
+  isBaseToQuote: false		// long
+  isExactInput: false		// need exact output amount of base token
+  amount: vBaseTradeSize
+  sqrtPriceLimitX96: 0		// slippage protection
+}
+  
+```
++ *vBaseCostReal = Quoter.swap(swapParams)* // Includes slippage and protocol fees
++ *slippageCost = vBaseCostReal - vBaseCostIdeal*
+
+Calculate current leverage using AMM spot price
++ *vBasePositionValue = vBasePositionSize * spotPrice*
++ *currentLeverage = vBasePositionValue / (vBasePositionValue + vQuoteBalance + collateralBalance)*
+
+Calculate usdcAmountIn and set USDC externalPositionUnit such that DIM can use it for transfer calculation.
+The Perp account is a positive equity position
++ *owedRealizedPnLPositionUnit = (carriedOwedRealizedPnL + pendingFunding) / setToken.totalSupply*
++ *owedRealizedPnLDiscount = owedRealizedPnLPositionUnit * mintQuantity*
++ *usdcAmountIn = (vBaseCostReal / current leverage) + slippageCost + owedRealizedPnLDiscount*
++ *setToken.externalPositionUnit = usdcAmountIn / mintQuantity*
+
+</td></tr>
+<tr><td valign="top">componentIssuanceHook</td>
+<td>  
+
+Invoked as an externalPositionHook duuring DIM.resolveEquityPositions. Happens after DIM has read external USDC position (set during moduleIssuanceHook) and transferred in:
++ *usdcAmount = mintQuantity * setToken.externalPositionUnit*
+
+Deposit USDC from SetToken into PerpV2
++ *SetToken.invokeApprove(USDC, PerpV2Vault, usdcAmount)*  
++ *SetToken.invokeDeposit(PerpV2Vault, USDC, usdcAmount)*
+
+Open position  
+```solidity
+swapParams = {
+  baseToken: vBaseToken 
+  isBaseToQuote: false		// long
+  isExactInput: true		
+  amount: usdcAmount
+  sqrtPriceLimitX96: 0		// slippage protection
+}
+```
+  
++ *PerpV2ClearingHouse.openPosition(swapParams)*
+  
+</td>
+</tr>
+</table>
+
+
+
+## Redemption
+
+This section includes:
++ A flow chart and user story for generalized PerpModule with protocol specific logic delegated to a PerpAdapter
++ A detailed walkthrough of redemption logic that would be implemented for the PerpV2 protocol
+
+![](../assets/stip-005/perp_userflow_redeem.png "")
+
+### User Story: Redeeming 10 Sets 
+
+UserA owns 10 SetTokens which represent a leveraged Perp position in ETH and wants to redeem them for the Perp’s collateral token, USDC.  
+
+1. User calls *DIM.redeem(setToken, 10)*
+
+2. DIM calls *PerpModule moduleRedeemHook(setToken, 10)* which passes call through to PerpAdapter contract for protocol 
+   specific implementation.
+
+3. PerpAdapter:
+     + Reads vETH position size, margin debt, and collateral positions from protocol, calculating their position units
+     + Calculates amount of pending payments (funding, carried PnL) which existing set holders must settle as a position unit
+
+4. PerpAdapter calls PerpProtocol to sell *vETHPositionUnit * 10* amount (see Trade user flow for details). 
```

Comment:
Think you know this based on our convo but the SetToken needs to make the call here. Are we executing the trade here for gas reasons? Ideally I think the trade is executed in the componentIssuanceHook alongside the `withdraw`.